### PR TITLE
Add get group role map

### PR DESCRIPTION
**Pull request by gene1wood**

### CODE REVIEW

File: cloudformation/functions/get_group_role_map.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
+import boto3
+
+
+def get_paginated_results(product, action, key, credentials=None, args=None):
+    """Paginate through AWS API responses, combining them into a list
+
+    :param str product: AWS product name
```
[the remark this answers is not in the document]
Comment (april) [reply cloudformation/functions/get_group_role_map.py]: Also add `-> list`

File: cloudformation/functions/get_group_role_map.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    :param dict credentials: Optional AWS API credentials
+    :param dict args: Optional additional arguments to pass to action method
+    :return: list of responses from all pages
+    """
+
+    args = {} if args is None else args
+    credentials = {} if credentials is None else credentials
+    return [y for sublist in [x[key] for x in boto3.client(
+        product, **credentials).get_paginator(action).paginate(**args)]
+            for y in sublist]
```
Comment (april) [inline cloudformation/functions/get_group_role_map.py]: I am always in favor of clever code, but I'm not sure `x`, `y`, `sublist`, keyword expansion, and double list comprehensions makes for particularly maintainable code.

File: cloudformation/functions/get_group_role_map.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    in that policy document. Return a map that looks like
+
+    {
+        'team_bar': [
+            'arn:aws:iam::123456789012:role/role-for-anyone-but-team-bar',
+            'arn:aws:iam::123456789012:role/project-baz-role'
+        ],
+        'team_foo': [
+            'arn:aws:iam::123456789012:role/role-for-team-foo',
+            'arn:aws:iam::123456789012:role/project-baz-role']
```
Comment (april) [inline cloudformation/functions/get_group_role_map.py]: Fix `]` to next line.

File: cloudformation/functions/get_group_role_map.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    :return: map of IAM ARNs to related OIDC claimed group names
+    """
+    assumed_role_credentials = {}
+    role_group_map = {}
+    for assumed_role_arn in assumed_role_arns:
+        aws_account_id = assumed_role_arn.split(':')[4]
+        client_sts = boto3.client('sts')
+        response = client_sts.assume_role(
+            RoleArn=assumed_role_arn,
+            RoleSessionName='Federated-Login-Policy-Collector')
```
Comment (april) [inline cloudformation/functions/get_group_role_map.py]: I think the `)` should be on the next line.

File: cloudformation/functions/get_group_role_map.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        assumed_role_credentials[aws_account_id] = {
+            'aws_access_key_id': response['Credentials']['AccessKeyId'],
+            'aws_secret_access_key': response['Credentials']['SecretAccessKey'],
+            'aws_session_token': response['Credentials']['SessionToken']}
```
Comment (april) [inline cloudformation/functions/get_group_role_map.py]: I'd move `}` to next line here too, although I'm honestly not sure which is the PEP8 way.

File: cloudformation/functions/get_group_role_map.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        roles = get_paginated_results(
+            'iam', 'list_roles', 'Roles', assumed_role_credentials[aws_account_id])
+        for role in roles:
+            groups = []
+            groups.extend(get_federated_groups_for_policy(
+                role['AssumeRolePolicyDocument']))
+            role_group_map[role['Arn']] = groups
```
Comment (april) [inline cloudformation/functions/get_group_role_map.py]: Is there a reason why this isn't just `groups = get_federated_groups_for_policy(role['AssumeRolePolicyDocument'])`? Why the `extend` on an empty list?

File: cloudformation/functions/get_group_role_map.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    """Paginate through AWS API responses, combining them into a list
+
+    :param str product: AWS product name
+    :param str action: AWS API action name
+    :param str key: The parent key in the paginated response
+    :param dict credentials: Optional AWS API credentials
+    :param dict args: Optional additional arguments to pass to action method
+    :return: list of responses from all pages
+    """
+
```
[the remark this answers is not in the document]
Comment (gene1wood) [reply cloudformation/functions/get_group_role_map.py]: Can you tell me more, I'm not seeing what you mean.

File: cloudformation/functions/get_group_role_map.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+def get_paginated_results(product, action, key, credentials=None, args=None):
+    """Paginate through AWS API responses, combining them into a list
+
+    :param str product: AWS product name
+    :param str action: AWS API action name
+    :param str key: The parent key in the paginated response
+    :param dict credentials: Optional AWS API credentials
+    :param dict args: Optional additional arguments to pass to action method
```
[the remark this answers is not in the document]
Comment (gene1wood) [reply cloudformation/functions/get_group_role_map.py]: Good idea, done.

File: cloudformation/functions/get_group_role_map.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (gene1wood) [reply cloudformation/functions/get_group_role_map.py]: Sounds good. I added the type hints but left the docstring content in to conform to PEP 257

File: cloudformation/functions/get_group_role_map.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+    return [y for sublist in [x[key] for x in boto3.client(
+        product, **credentials).get_paginator(action).paginate(**args)]
+            for y in sublist]
+
+
+def flip_map(arn_group_map):
+    """Flip the map of ARN to group list to group to ARN list
+
+    :param dict arn_group_map: ARN to group list map
+    :return: Group to ARN list map
```
[the remark this answers is not in the document]
Comment (gene1wood) [reply cloudformation/functions/get_group_role_map.py]: Ah much cleaner, done

File: cloudformation/functions/get_group_role_map.py (hunk 2 above)
Comment (gene1wood) [reply cloudformation/functions/get_group_role_map.py]: I expanded this and improved the variable names

File: cloudformation/functions/get_group_role_map.py (hunk 4 above)
Comment (gene1wood) [reply cloudformation/functions/get_group_role_map.py]: Yup, fixed with black

File: cloudformation/functions/get_group_role_map.py (hunk 6 above)
Comment (gene1wood) [reply cloudformation/functions/get_group_role_map.py]: Good catch. This was an artifact from a previous approach I was taking. Fixed now

File: cloudformation/functions/get_group_role_map.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (april) [reply cloudformation/functions/get_group_role_map.py]: This function has a space between the docstring and the code (which I prefer) and the rest of them do not.  :)

[the remark this answers is not in the document]
Comment (gene1wood) [reply cloudformation/functions/get_group_role_map.py]: Ahh got it. Ok, I removed the space (to conform to PEP257)

**Pull request merged**